### ISSUE TITLE
The need for an injected services setup specific to the design time environment

### ISSUE THREAD

**Issue by ApolonTorq** [radzenhq/radzen-blazor-studio#32]
I'm working on a blazor app that is being built from an existing domain/data/services layer. Once RBS was successful in opening and compiling the application, the next issue I experienced was the practicality of using the design surface with an existing set of complex application injected services.

When running in the design surface, none of the injected services are available which results in "There is no registered service" exceptions. It seems there is also no way of making the [Inject] attribute make the service optional. Though a [workaround for this](https://stackoverflow.com/a/62291903) exists, it would bloat the code and also force the injection of IServiceProvider every where which is "ugly as".

Even if injected services could be null when running in the RBS design surface, the application would then need to be littered with null checks to change the component's behaviour when running in the inferred design surface state. This does remind me of the design mode checks that can be made for WPF and Windows Forms designer specific code. Even additional design mode check functionality, though useful, does "pollute" the component application code with functionality not related to its core purpose.

To resolve this "real world" development issue for using RBS, would you consider adding the ability for application programmers to specify the service configuration to be used at RBS design time? It would allow the application programmer to tailor the design surface behaviour to their application. This would be particularly useful when RBS is being used with an existing data/services layer. I believe it would also allow RBS to be used in many more Blazor application development scenarios in practice (as opposed to in theory).

One way this could be implemented is:

- There is a Radzen.Design.dll or Radzen.ComponentModel.dll containing design time support code for RBS.
- A class or static function can be marked with a [DesignTimeSetup] attribute to mark it as defining the code needed to set up the runtime services needed by the design time environment.
- When working with a solution, RBS looks for the [DesignTimeSetup] attribute across the projects and runs that code to initialize the design time environment and services scope.
- The application programmer can then create a DLL project solely for their design time setup e.g. MyApp.DesignTime.csproj. That cs project can have dependencies on the data/services layers common to the actual application. The actual application would not have a dependency on the MyApp.DesignTime project but that design time project would be in the solution to provide instructions to RBS on how to get a functional and useful design time environment for the application programmer.

This implementation is just an example approach. Any approach or variation that solves the issue of requiring design time specific services injected would be welcome.

**Comment by akorchev** [comment]
RBS is supposed to inject all services registered in a "known" location such as a Startup class or Program.cs. Thus all [Inject] properties should be set by design. Where does your application register those services? Perhaps RBS doesn't find them at the moment. As usual you could send us an app which we can use to troubleshoot. Thanks!

**Comment by ApolonTorq** [comment]
The solution has a Program.cs and a Startup.cs file. Its going to be difficult to send cut-down app source code including these files because of the dependency explosion. I'd rather get more info from RBS on what it did looking for Program/Startup patterns and what it found or rejected. Does it log its attempts with respect to service registration?

**Comment by akorchev** [comment]
RBS looks for ConfigureServices method in a class called Startup. If it fails to find one it looks for the Program (or a class with a static Main method). Currently there isn't any logging involved.

**Comment by ApolonTorq** [comment]
When it finds ConfigureServices method in a class called Startup, does it just run that method in isolation?
Is the goal just to register services and nothing else?

If that's the case, it may be finding the Setup class, but not running the AutofacServiceProviderFactory related code which registers a bunch of services using Autofac rather than Microsoft dependency injection.

**Comment by akorchev** [comment]
> but not running the AutofacServiceProviderFactory related code which 

It runs only the ConfigureServices method if it has one. How does one use Autofac in a Blazor application? I am not familiar with it.

**Comment by ApolonTorq** [comment]
There's an adapter applied to the host builder as per [this example](https://github.com/autofac/Examples/blob/master/src/GenericHostBuilderExample/Program.cs).

There's going to be a wide variety of startup coding approaches "out in the wild". Rather than taking that on, how about having a convention which allows a tailored RBS specific ConfigureServices to be run?

e.g. if RBS finds a ConfigureDesignTimeServices function it uses that in preference to ConfigureServices.

It would be a "way out" for any setup/dependency-injection scenario that isn't the Blazor default. It would also be an optional mechanism for taking control of the design time service environment if need be.

**Comment by akorchev** [comment]
> e.g. if RBS finds a ConfigureDesignTimeServices function it uses that in preference to ConfigureServices.

We can probably do that however without an actual application it would be very hard to test and do properly. Can you create a minimal Blazor app that uses Autofac? It should probably fail in RBS in the same way as your actual application.

**Comment by ApolonTorq** [comment]
Ok, will do (guessing within a week).

**Comment by ApolonTorq** [comment]
Ok, the sample project that demonstrates the issue is attached here. The README.md is also reproduced here:

### Sample Code Description

This sample Blazor application illustrates a real world scenario where a new Blazor app is being developed which uses an existing .NET library based on Autofac as the dependency injection framework instead of the Microsoft DI functionality.

The Blazor specific services are still injected in `Startup.ConfigureServices`. However, all of the external library services are configured using the Autofac API. This scenario is simulated in this sample with the Autofac service registration being performed in `Program.ConfigureAutofac`.

As it stands, this sample application:

- Builds and runs correctly. Success is defined by successfully running the /fetchdata page and seeing a temperature table because the WeatherForecastService is resolved by Autofac.
- RBS Design mode fails. To reproduce, attempt to edit the FetchData.razor page to see a rendering exception which stops the user from being able to edit that page.
    - `Cannot provide a value for property 'ForecastService' on type 'RadzenStudioAutofac.Pages.FetchData'. There is no registered service of type 'RadzenStudioAutofac.Data.WeatherForecastService'.`

The proposed solution to this issue is to add support for running the `Startup.ConfigureDesignTimeServices` function to initialize services specifically for the RBS design mode. This sample code has such a function already in place to be able to test the proposed fix.

Note that the `ConfigureDesignTimeServices` function support allows app developers to also provide custom service registration suitable to RBS design mode.

[RadzenStudioAutofac.zip](https://github.com/radzenhq/radzen-blazor-studio/files/8683903/RadzenStudioAutofac.zip)

**Comment by akorchev** [comment]
Do you think leaving ConfigureDesignTimeServices is feasible like this?

```
    public void ConfigureDesignTimeServices(IServiceCollection services)
    {
        services.AddSingleton<WeatherForecastService>();
    }
```

We would like to call ConfigureServices always (if available).

**Comment by ApolonTorq** [comment]
Yes, but can you call ConfigureDesignTimeServices first by convention?
It would allow the Startup object to effectively be informed its running within a design time environment and change its behaviour if needed.

**Comment by akorchev** [comment]
We released 1.0.0.preview.13 which will invoke a method called ConfigureDesignTimeServices in the Startup class (before ConfigureServices).

```
    public void ConfigureDesignTimeServices(IServiceCollection services)
    {
        services.AddSingleton<WeatherForecastService>();
    }
```
